### PR TITLE
Feature/car selection

### PR DESCRIPTION
**Pull request by m4rtint**
IMPORTANT PART - InputManager

### CODE REVIEW

File: Assets/Script/Data/GameData.cs [hunk 1]
```diff
@@ -12,10 +12,10 @@
 public static class GameData
 {
     private static int m_playerCount = -1;
-    private static List<PlayerInformation> m_playerInformationList = new List<PlayerInformation>()
+    private static Dictionary<Player, PlayerInformation> m_playerInformationList = new Dictionary<Player, PlayerInformation>
     {
-        new PlayerInformation(Player.Player_1, new PlayerDataLoader(), new PlayerControllerInformation(PCControl.WASD)),
-        new PlayerInformation(Player.Player_2, new PlayerDataLoader(), new PlayerControllerInformation(PCControl.KEYS))
+        {Player.Player_1, new PlayerInformation(Player.Player_1, 0, new PlayerDataLoader(), new PlayerControllerInformation(PCControl.WASD)) },
```
Comment (m4rtint) [inline Assets/Script/Data/GameData.cs]: Changes to the playerinformation class

File: Assets/Script/Data/PlayerDataLoader.cs [hunk 2]
```diff
@@ -72,6 +66,25 @@ public Color PlayerColor(Player player)
...
+        int amountOfCars = m_carSprites.Length;
+
+        this.m_AllCars = new CarObj[amountOfCars];
+        for (int i = 0; i < amountOfCars; i++)
+        {
+            m_AllCars[i] = new CarObj(m_carSprites[i]);
+        }
+    }
+
+    public struct CarObj {
```
Comment (m4rtint) [inline Assets/Script/Data/PlayerDataLoader.cs]: I was planning for the player choose a CarObj - containing player's color and car sprite

BUT - problems happened, but leaving this here for now. 
Details in Utilities class

File: Assets/Script/Data/PlayerInformation.cs [hunk 3]
```diff
@@ -27,13 +27,13 @@ public class PlayerInformation
...
         /// <param name="controllerInfo">Players controller input information</param>
-        public PlayerInformation(Player number, PlayerDataLoader loader, PlayerControllerInformation controllerInfo)
+        public PlayerInformation(Player number, int pointer, PlayerDataLoader loader, PlayerControllerInformation controllerInfo)
         {
             this.m_playerNumber = number;
             this.m_loader = loader;
             this.m_playerControllerInformation = controllerInfo;
             this.m_color = loader.PlayerColor(number);
-            this.m_sprite = loader.PlayerCarSprite(number);
+            this.m_sprite = loader.PlayerCarSprite(pointer);
```
Comment (m4rtint) [inline Assets/Script/Data/PlayerInformation.cs]: loader.PlayerCarSprite => array of car sprites.

the pointer, just points to which index of car sprites

File: Assets/Script/Data/Utilities.cs [hunk 4]
```diff
@@ -28,5 +28,50 @@ public static Color ColorFrom255(float r, float g, float b, float alpha = m_Full
         {
             return new Color(r / m_FullValue, g / m_FullValue, b / m_FullValue, alpha / m_FullValue);
         }
+
+        public static Color AverageColor(Texture2D texture)
```
Comment (m4rtint) [inline Assets/Script/Data/Utilities.cs]: Was planning to either get average sprite color OR dominant color.

BUT problems:
it takes the WHOLE car sprite sheet which contains every single car. 
I already cut up the car sprite sheet, and passed in a single car sprite, but the "texture" is still the WHOLE car sprite.

These functions are useless for now, but will be used in the future..maybe.

File: Assets/Script/Managers/InputManager.cs [hunk 5]
```diff
@@ -16,6 +16,43 @@ namespace AngryGooseGames
     /// </summary>
     public static class InputManager
     {
+
```
Comment (m4rtint) [inline Assets/Script/Managers/InputManager.cs]: I rearranged the methods around. 
Put regions in this.

AllTypes (checks all input method presses)
PC exclusive
Controller exclusive

File: Assets/Script/Managers/InputManager.cs [hunk 6]
```diff
@@ -55,46 +92,92 @@ public static bool IsBreaking(PlayerControllerInformation inputInfo)
...
-        /// Pause or Resume the game
-        /// </summary>
-        public static void PauseAndPlay()
+        private static bool IsPCControlBreakPressed(PCControl control)
         {
-            if (Input.GetButtonDown("Pause") || IsStartPressed())
+            return Input.GetButtonDown(control + "_Break");
+        }
+
+        public static float IsPCControlHorizontalPressed(PCControl control)
```
Comment (m4rtint) [inline Assets/Script/Managers/InputManager.cs]: @mikeslgoh @wbeltran777 @Gursaanj 
Off the top of my head, the only way I can figure this out. 

 When you do get axis, it slowly goes from 0 -> 1/-1.

So car selection would be called many many times. since it checks (isHorizontalPressed > 1)


This solution is to basically detect a button/key pressed ONCE and ONLY ONCE per frame


If you have any better solution please let me know

File: Assets/Script/UI/ControllerSelection.cs [hunk 7]
```diff
@@ -60,45 +72,77 @@ private void AddControllersIfNeeded()
...
-        m_amountOfPlayersPlaying++;
+        m_playerInfoList.Add(player, new PlayerInformation(player, pointer, m_playerDataLoader, info));
         UpdateGameData();
+        UpdateReadyAmountOfPlayers(true);
+        return true;
+    }
+
+    private void UpdateReadyButtonIfNeeded()
+    {
+        m_nextButton.interactable = m_amountOfReadyPlayers == m_amountOfPlayersPlaying;
```
Comment (m4rtint) [inline Assets/Script/UI/ControllerSelection.cs]: the "ready" button is only activated when everyone is ready

File: Assets/Script/UI/ControllerSelectionPanel.cs [hunk 8]
```diff
@@ -3,33 +3,288 @@
...
+        m_border.color = m_loader.AllCars[m_currentPointer].m_Color;
+        m_carSprite.sprite = m_loader.AllCars[m_currentPointer].m_Sprite;
+    }
+
+    private int NumberOfCars()
+    {
+        return m_loader.AllCars.Length;
+    }
+
+    private void CheckDependencies()
```
Comment (m4rtint) [inline Assets/Script/UI/ControllerSelectionPanel.cs]: Dependency checks.  let's see if it's easier to understand any broken builds.

File: Assets/Script/Data/Utilities.cs (hunk 4 above)
Comment (mikeslgoh) [reply Assets/Script/Data/Utilities.cs]: Do you mean that you cut up the car sprite sheet outside of Unity and it still uses the WHOLE sheet with 4 cars?

Comment (m4rtint) [reply Assets/Script/Data/Utilities.cs]: Unity had a function.
Milticut sprite

Where you cut out a sprite sheet into individual sprites

Comment (m4rtint) [reply Assets/Script/Data/Utilities.cs]: If you go to Th resources folder under sprites, you'll get what I mean.

And yes it still takes the whole sprite with 9 cars

File: Assets/Script/Managers/InputManager.cs (hunk 6 above)
Comment (mikeslgoh) [reply Assets/Script/Managers/InputManager.cs]: I think once key press is the way to go. Do we have to put the check there? Because I thought for Key/Button down, it would only work when it is pressed and that's all

File: Assets/Script/UI/ControllerSelection.cs (hunk 7 above)
Comment (mikeslgoh) [reply Assets/Script/UI/ControllerSelection.cs]: So it only detects if it is waiting for a player if they are on the car selection right? It doesn't matter if a controller is plugged in as long as they don't join the game?

File: Assets/Script/UI/ControllerSelectionPanel.cs (hunk 8 above)
Comment (mikeslgoh) [reply Assets/Script/UI/ControllerSelectionPanel.cs]: Let's see how this works and then maybe consider adding it in other areas of our code base.

File: Assets/Script/Managers/InputManager.cs (hunk 6 above)
Comment (m4rtint) [reply Assets/Script/Managers/InputManager.cs]: Yeah exactly. I'm checking if it's pressed once. Not anymore.

File: Assets/Script/UI/ControllerSelection.cs (hunk 7 above)
Comment (m4rtint) [reply Assets/Script/UI/ControllerSelection.cs]: Exactly.

File: Assets/Script/Data/Utilities.cs (hunk 4 above)
Comment (mikeslgoh) [reply Assets/Script/Data/Utilities.cs]: maybe we should consider trying to split it outside of Unity?

File: Assets/Script/Managers/InputManager.cs (hunk 6 above)
Comment (mikeslgoh) [reply Assets/Script/Managers/InputManager.cs]: ok yeah I think this is the best approach

File: Assets/Script/Data/Utilities.cs (hunk 4 above)
Comment (m4rtint) [reply Assets/Script/Data/Utilities.cs]: I think that's the only way.

Comment (m4rtint) [reply Assets/Script/Data/Utilities.cs]: I'll probably do this afterwards

File: Assets/Script/UI/ControllerSelection.cs [hunk 9]
```diff
@@ -40,11 +42,21 @@ public class ControllerSelection : MonoBehaviour
...
+    private void Start()
+    {
+        foreach(ControllerSelectionPanel panel in m_players)
+        {
+            panel.OnCarSelected += AddPlayerInfo;
+            panel.OnRemoveSelectedCar += RemovePlayerInfo;
+        }
+    }
+
     private void Update()
```
Comment (wbeltran777) [inline Assets/Script/UI/ControllerSelection.cs]: Should we be subscribing this to the global update in world manager?

Comment (m4rtint) [reply Assets/Script/UI/ControllerSelection.cs]: WorldManager is not in this menu scene yet.

And usually we have a worldmanager global update to pause/play/slow down time for stuff, but menu don't have that stuff.  - that's what i'm presuming

Comment (mikeslgoh) [reply Assets/Script/UI/ControllerSelection.cs]: I think we have a task to do a UIDelegate I believe

Comment (m4rtint) [reply Assets/Script/UI/ControllerSelection.cs]: If i remember correctly, the UI Delegate is more so to update the score on the game scenes

Comment (mikeslgoh) [reply Assets/Script/UI/ControllerSelection.cs]: ah ok, that's probably it. Do we need a menu/panel update then? For consistency?

Comment (m4rtint) [reply Assets/Script/UI/ControllerSelection.cs]: So for the worldmanager is good to do it so that you can pause/play/slow down time from ONE place only.

If there's a reason for the main menu, sure. But I can't think of a reason 

The main usage of the update in menu is to detect user input. 
Only reasoning I can think of, is to "disable user input".

Comment (mikeslgoh) [reply Assets/Script/UI/ControllerSelection.cs]: oh sorry. Uhh I say that if there is no need to make this change except for just consistency, then it's fine to leave it as it is.

@Gursaanj @wbeltran777 ?

File: Assets/Script/UI/ControllerSelectionPanel.cs [hunk 10]
```diff
@@ -3,33 +3,288 @@
...
+        {
+            ScrollThroughCars(true);
+        }
+        else if (m_inputDevice.LeftStickRight.WasPressed)
+        {
+            ScrollThroughCars(false);
+        }
+        else if (m_inputDevice.Action1.WasPressed)
+        {
+            CarSelected(m_inputDevice, m_playerNumber, m_currentPointer);
```
Comment (wbeltran777) [inline Assets/Script/UI/ControllerSelectionPanel.cs]: Small bug, if a player presses 'down' as their method of connecting their controller, then they'll also select the first car on the sprite list as long as the car is not taken. This might cause slight confusion for showcasing as the first player to join may sometimes be set to READY right away and may be confused as to why they weren't able to pick a car.

File: Assets/Script/UI/ControllerSelection.cs (hunk 9 above)
Comment (wbeltran777) [reply Assets/Script/UI/ControllerSelection.cs]: Stuff seems to be working fine, so if we don't see any long term damage to not doing a global manager for UI, I say we don't worry about it.

**Comment by m4rtint** [comment]
[Player Selection - Choose Cars](https://trello.com/c/Z0FJPPPq/212-player-selection-choose-cars)

File: Assets/Script/Managers/InputManager.cs (hunk 6 above)
Comment (Gursaanj) [reply Assets/Script/Managers/InputManager.cs]: Yup I thinks thats a good approach

File: Assets/Script/UI/ControllerSelection.cs (hunk 9 above)
Comment (Gursaanj) [reply Assets/Script/UI/ControllerSelection.cs]: Fair enough, Im down either way. but if this is the only case where update is begin called of the menu scene, then I guess we dont have to worry about it

File: Assets/Script/UI/ControllerSelectionPanel.cs (hunk 10 above)
Comment (m4rtint) [reply Assets/Script/UI/ControllerSelectionPanel.cs]: Lol that ain't a small bug. 

I think that part will be solved, when we figure out UX wise, how to "go next" .

Cuz I think we're thinking of pressing start to go next, instead of press "down" / "Accerlate"/"confirm"

File: Assets/Script/UI/ControllerSelection.cs (hunk 9 above)
Comment (m4rtint) [reply Assets/Script/UI/ControllerSelection.cs]: I think @wbeltran777 has a point. 

I'm thinking that we'll probably implement animated transitions between menu scenes.. And during the transitions, the players should not be able to press any buttons. 

High chance to have a global main menu delegate.

Comment (m4rtint) [reply Assets/Script/UI/ControllerSelection.cs]: But that's for another PR

File: Assets/Script/UI/ControllerSelectionPanel.cs (hunk 10 above)
Comment (wbeltran777) [reply Assets/Script/UI/ControllerSelectionPanel.cs]: Okay gotcha! In that case things look pretty great!

Comment (m4rtint) [reply Assets/Script/UI/ControllerSelectionPanel.cs]: Oh wait, i think i misunderstood what you meant. Now I understand. 

The InControl framework handles i nicely. it only detects the "OnButttonPressed" ONCE, so it won't detect it more than once.

Comment (wbeltran777) [reply Assets/Script/UI/ControllerSelectionPanel.cs]: Right, but it's still triggering both, sync controller, and "Ready" at the same time, if you press 'S' or 
'Down Arrow' to join. Is this an issue we should be concerned about?

Comment (m4rtint) [reply Assets/Script/UI/ControllerSelectionPanel.cs]: Yes, I'll fix it in this branch

Comment (m4rtint) [reply Assets/Script/UI/ControllerSelectionPanel.cs]: @wbeltran777 
Fixed it, not sure if it's a work around or fix but..changed ControllerSelection to use 
LateUpdate, and ControllerSelectionPanel still uses Update.

Comment (m4rtint) [reply Assets/Script/UI/ControllerSelectionPanel.cs]: yeah just googled, this is not a good fix. But if it's merged. it works lol. Will think of a better solution soon.

**Comment by mikeslgoh** [comment]
I'll approve once we adjust the look to match the mocks (which seems to be pretty straight forward). Other than that, everything else looks good :)

**Comment by m4rtint** [comment]
@mikeslgoh 
I kinda wanna approve this first..and then the looks in another PR. 

this PR is big enough as it is.

**Comment by mikeslgoh** [comment]
[Change way for players to join the match](https://trello.com/c/QiQIcgzi/206-change-way-for-players-to-join-the-match)

**Comment by mikeslgoh** [comment]
[Change way for players to join the match](https://trello.com/c/QiQIcgzi/206-change-way-for-players-to-join-the-match)
